### PR TITLE
[stdlib] Set, Dictionary, AnyHashable: __consuming/__owned audit

### PR DESCRIPTION
**Pull request by lorentey**
Add `__consuming` and `__owned` to Set and Dictionary members where applicable, including internal implementation details.

Ignore compiler intrinsics for casting for now — their ARC behavior is covered by unit tests that need to be updated.

rdar://problem/44725231

### CODE REVIEW

**Comment by lorentey** [comment]
@swift-ci please test

**Comment by lorentey** [comment]
Commit f59fd89 also audits SetAlgebra, and removes `__owned` from the first parameter of intersection and subtraction methods.

**Comment by lorentey** [comment]
This did not have much effect on performance in my own benchmarking, but let's see an official verdict:

@swift-ci please smoke benchmark

**Comment by lorentey** [comment]
Nice; I guess insertions did get a bit faster in some contexts?

**Comment by lorentey** [comment]
```
01:15:15.486 --- /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/test/SourceKit/DocSupport/doc_clang_module.swift.response	2018-09-24 06:33:53.000000000 -0700
01:15:15.486 +++ /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/SourceKit/DocSupport/Output/doc_clang_module.swift.tmp.response	2018-09-24 07:47:26.000000000 -0700
01:15:15.486 @@ -61,7 +61,7 @@
01:15:15.486  
01:15:15.486  extension FooRuncingOptions {
01:15:15.486  
01:15:15.486 -    @inlinable init<S>(_ sequence: S) where S : Sequence, FooRuncingOptions.Element == S.Element
01:15:15.486 +    @inlinable init<S>(_ sequence: __owned S) where S : Sequence, FooRuncingOptions.Element == S.Element
``

I didn't know `__owned` in protocol requirements get automatically applied on conformances, but it makes perfect sense now!

File: stdlib/public/core/Set.swift [hunk 1]
```diff
@@ -732,8 +735,8 @@ extension Set: SetAlgebra {
   /// - Returns: `true` if the set is a subset of `possibleSuperset`;
   ///   otherwise, `false`.
   @inlinable
-  public func isSubset<S: Sequence>(of possibleSuperset: S) -> Bool
-    where S.Element == Element {
+  public func isSubset<S: Sequence>(of possibleSuperset: __owned S) -> Bool
```
[the remark this answers is not in the document]
Comment (lorentey) [reply stdlib/public/core/Set.swift]: Oops, no, this slipped in by mistake. `SetAlgebra` defines a similar extension with a non-owned sequence parameter.

I think this makes sense assuming we'll eventually provide a non-consuming way to iterate over the borrowed contents of a sequence. In fact, `Sequence` has a non-consuming `forEach`, so (if that's not a bug) this is technically already possible. ¯\_(ツ)_/¯

**Comment by lorentey** [comment]
@swift-ci please test

File: stdlib/public/core/Set.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (airspeedswift) [reply stdlib/public/core/Set.swift]: > In fact, `Sequence` has a non-consuming `forEach`

LOL

File: stdlib/public/core/Array.swift [hunk 2]
```diff
@@ -1784,7 +1784,7 @@ extension Array {
 
 extension Array: _HasCustomAnyHashableRepresentation
   where Element: Hashable {
-  public func _toCustomAnyHashable() -> AnyHashable? {
+  __consuming public func _toCustomAnyHashable() -> AnyHashable? {
```
Comment (airspeedswift) [inline stdlib/public/core/Array.swift]: I think I found when updating tests that process these getting spat back out that they come out as `public __consuming func` so I've done the others that way around.

Comment (airspeedswift) [reply stdlib/public/core/Array.swift]: ah in fact this is just the odd-one out

**Comment by lorentey** [comment]
@swift-ci please test and merge

**Comment by lorentey** [comment]
@swift-ci please test and merge

**Pull request merged**